returnn/tf/layers/base.py

```diff
@@ -491,18 +492,26 @@ def transform_config_dict(cls, d, network, get_layer):
           if target.startswith("layer:"):
             target_layers[target] = get_layer(target[len("layer:"):])
           else:
-            # Note: This is a workaround for cases where we need to know about used data keys before the layer
-            # itself is constructed (e.g. in _SubnetworkRecCell.get_output).
-            # A nicer solution would be to not modify this here,
-            # but instead lazily handle it in TFNetwork.get_extern_data,
-            # such that we do not need to know in advance which data keys we need.
-            # Also, if we are inside a rec layer, and doing search, we also cannot do that.
-            if network.is_inside_rec_layer() and not network.search_flag:
-              network.get_extern_data(target, mark_data_key_as_used=True)
-            if not network.search_flag:
-              # Also, there are cases when we want to have the target as an explicit layer dep,
-              # e.g. when the target has a beam, to derive the search choices.
-              target_layers[target] = get_layer("data:%s" % target)
+            try:
+              # Note: This is a workaround for cases where we need to know about used data keys before the layer
+              # itself is constructed (e.g. in _SubnetworkRecCell.get_output).
+              # A nicer solution would be to not modify this here,
+              # but instead lazily handle it in TFNetwork.get_extern_data,
+              # such that we do not need to know in advance which data keys we need.
+              # Also, if we are inside a rec layer, and doing search, we also cannot do that.
+              if network.is_inside_rec_layer() and not network.search_flag:
+                network.get_extern_data(target, mark_data_key_as_used=True)
+              else:
+                # Anyway, to check for the name.
+                network.get_extern_data(target, mark_data_key_as_used=False)
+            except DataNotFound:
+              # Expect that this is a layer.
+              target_layers[target] = get_layer(target)
+            else:
+              if not network.search_flag:
```

Comment:
Actually I don't understand the existing code here. Is it `not network.search_flag` because in the search case `ChoiceLayer` would add the dependency? If so, this is already a pretty bad. 😅

Reply:
I don't exactly remember anymore. I think it was just buggy, in a way that this would trigger `get_extern_data(target, mark_data_key_as_used=True)`, and then the rec layer would try to unroll the targets in the loop, which would not work because the loop can potentially run different number of steps than the target (in search mode).

Or maybe sth similar/related. You are right in principle. Or actually, it's not really well defined what the targets should be when search mode is enabled. It cannot unroll it. It can still provide the template (with `dim` and maybe `vocab`, but ignoring time-dim). So maybe that should be it, but then as template only (maybe using `DataNotAvailableLayer`).

I think when I wrote this code, I was just too lazy to really think about how it should be for search mode, and just implemented this case.

But anyway, I think solving this would be better for a separate PR.